In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
!python -m pip install oaklib

In [ ]:
# node.js needs to be at least version 12 or so for OAK
!node --version

In [2]:
# Files
old_scores = 'https://kg-hub.berkeleybop.io/frozen_incoming_data/hp-mp-phenodigm-cache.txt.gz'
new_scores = 'https://kg-hub.berkeleybop.io/kg-phenio/hp-mp_similarities.tar.gz'

old_tar_file = 'hp-mp-phenodigm-cache.txt.gz'
new_tar_file = 'hp-mp_similarities.tar.gz'

old_sim_file = 'hp-mp-phenodigm-cache.txt'
new_sim_file = 'KGPhenio_similarities'

In [ ]:
!wget {old_scores}
!gunzip {old_tar_file}

!wget {new_scores}
!tar -xvzf {new_tar_file}

In [ ]:
!ls

In [3]:
old_sim_pd = pd.read_csv(old_sim_file, sep="\t", header=None)
mapping = {old_sim_pd.columns[0]:'source_old',
           old_sim_pd.columns[1]: 'destination_old', 
           old_sim_pd.columns[2]: 'jaccard_old',            
           old_sim_pd.columns[3]: 'resnik_score_old',
           old_sim_pd.columns[4]: 'subsumer_old',
           }
old_sim_pd = old_sim_pd.rename(columns=mapping)

old_sim_pd['source_old'] = old_sim_pd['source_old'].str.replace('_',':')
old_sim_pd['destination_old'] = old_sim_pd['destination_old'].str.replace('_',':')

old_sim_pd

source_old destination_old  jaccard_old  resnik_score_old  \
0         HP:0000033      MP:0004014     0.470588          2.659134   
1         HP:0000033      MP:0020998     0.434783          3.300796   
2         HP:0000033      MP:0020957     0.363636          2.659134   
3         HP:0000033      MP:0020869     0.212766          3.300796   
4         HP:0000033      MP:0020824     0.500000          3.300796   
...              ...             ...          ...               ...   
13033730  HP:0008609      MP:0004176     0.243243          5.040392   
13033731  HP:0008609      MP:0030154     0.928571          7.560589   
13033732  HP:0008609      MP:0030155     0.866667          7.560589   
13033733  HP:0008609      MP:0030156     0.650000          7.560589   
13033734  HP:0008609      MP:0030158     0.433333          7.560589   

         subsumer_old  
0         MP_0002160;  
1         MP_0009198;  
2         MP_0002160;  
3         MP_0009198;  
4         MP_0009198;  
...               ...  
13033730  HP_0031703;  
13033731  MP_0000049;  
13033732  MP_0000049;  
13033733  MP_0000049;  
13033734  MP_0000049;  

[13033735 rows x 5 columns]

In [ ]:
old_sim_pd['jaccard_old'].plot(kind='hist', xlim=0)

In [ ]:
old_sim_pd['resnik_score_old'].plot(kind='hist', bins=15, xlim=0)

In [5]:
new_sim_pd = pd.read_csv(new_sim_file, sep=",")
new_sim_pd

source destination  resnik_score   jaccard
0         MP:3000005  MP:3000005     14.526168  1.000000
1         HP:0030350  HP:0030350     14.526168  1.000000
2         MP:0008925  MP:0008925     14.526168  1.000000
3         MP:0008926  MP:0008926     14.526168  1.000000
4         MP:0008926  MP:0013185     14.526168  0.857143
...              ...         ...           ...       ...
51383633  HP:0001802  HP:0200054      5.041991  0.214286
51383634  HP:0100142  MP:0009001      5.041991  0.117647
51383635  HP:0010388  MP:0009001      5.041991  0.133333
51383636  HP:0010323  HP:0001802      5.041991  0.214286
51383637  HP:0010374  MP:0009001      5.041991  0.214286

[51383638 rows x 4 columns]

In [ ]:
new_sim_pd['jaccard'].plot(kind='hist', xlim=0)

In [ ]:
new_sim_pd['resnik_score'].plot(kind='hist', bins=15, xlim=0)

In [6]:
intersected_scores1 = old_sim_pd.merge(new_sim_pd, how='inner',
                                      left_on=['source_old', 'destination_old'], right_on=['source', 'destination'])
intersected_scores2 = old_sim_pd.merge(new_sim_pd, how='inner',
                                      left_on=['source_old', 'destination_old'], right_on=['destination', 'source'])
intersected_scores = pd.concat([intersected_scores1, intersected_scores2])
intersected_scores

source_old destination_old  jaccard_old  resnik_score_old  \
0        HP:0000033      MP:0004014     0.470588          2.659134   
1        HP:0000033      MP:0020998     0.434783          3.300796   
2        HP:0000033      MP:0020957     0.363636          2.659134   
3        HP:0000033      MP:0020824     0.500000          3.300796   
4        HP:0000033      MP:0002771     0.400000          3.300796   
...             ...             ...          ...               ...   
2361470  HP:0008609      MP:0004204     0.500000          7.560589   
2361471  HP:0008609      MP:0030154     0.928571          7.560589   
2361472  HP:0008609      MP:0030155     0.866667          7.560589   
2361473  HP:0008609      MP:0030156     0.650000          7.560589   
2361474  HP:0008609      MP:0030158     0.433333          7.560589   

        subsumer_old      source destination  resnik_score   jaccard  
0        MP_0002160;  HP:0000033  MP:0004014      5.358735  0.153846  
1        MP_0009198;  HP:0000033  MP:0020998      7.041238  0.166667  
2        MP_0002160;  HP:0000033  MP:0020957      7.038994  0.153846  
3        MP_0009198;  HP:0000033  MP:0020824      6.481542  0.363636  
4        MP_0009198;  HP:0000033  MP:0002771      7.041238  0.153846  
...              ...         ...         ...           ...       ...  
2361470  MP_0000049;  MP:0004204  HP:0008609     10.195435  0.230769  
2361471  MP_0000049;  MP:0030154  HP:0008609     10.195435  0.272727  
2361472  MP_0000049;  MP:0030155  HP:0008609     10.195435  0.250000  
2361473  MP_0000049;  MP:0030156  HP:0008609     10.195435  0.272727  
2361474  MP_0000049;  MP:0030158  HP:0008609      8.446235  0.230769  

[5833600 rows x 9 columns]

In [ ]:
ax1 = intersected_scores.plot.scatter(x='resnik_score',
                                      y='resnik_score_old',
                                      c='DarkBlue',
                                      s=0.01
                                     )

In [ ]:
ax1 = intersected_scores.plot.scatter(x='jaccard',
                                      y='jaccard_old',
                                      c='DarkBlue',
                                      s=0.01
                                     )

### Which pairs have high scores in the old set and low scores in the new set?

In [7]:
intersected_scores['resnik_diff'] = intersected_scores['resnik_score_old'] - intersected_scores['resnik_score']
intersected_scores.sort_values(by='resnik_diff', ascending=False, inplace=True)
intersected_scores

source_old destination_old  jaccard_old  resnik_score_old  \
962688   HP:0002607      MP:0003281     1.000000         14.359005   
632818   HP:0040183      MP:0003281     0.944444         14.359005   
1327215  HP:0012700      MP:0003281     0.529412         14.351005   
1326917  HP:0012701      MP:0003281     0.500000         14.351005   
878712   HP:0012700      MP:0003281     0.529412         14.351005   
...             ...             ...          ...               ...   
1071039  HP:0410188      MP:0005562     0.282051          2.508834   
728809   HP:0410198      MP:0005562     0.321429          2.508834   
1069220  HP:0410198      MP:0005562     0.321429          2.508834   
2020913  HP:0002553      MP:0003761     0.222222          2.512377   
1361777  HP:0002553      MP:0003761     0.222222          2.512377   

        subsumer_old      source destination  resnik_score   jaccard  \
962688   MP_0003281;  HP:0002607  MP:0003281      5.159422  0.300000   
632818   MP_0003281;  HP:0040183  MP:0003281      5.159422  0.272727   
1327215  HP_0012700;  HP:0012700  MP:0003281      5.159422  0.333333   
1326917  HP_0012700;  HP:0012701  MP:0003281      5.159422  0.300000   
878712   HP_0012700;  MP:0003281  HP:0012700      5.159422  0.333333   
...              ...         ...         ...           ...       ...   
1071039  MP_0008469;  HP:0410188  MP:0005562     13.139874  0.300000   
728809   MP_0008469;  MP:0005562  HP:0410198     13.139874  0.300000   
1069220  MP_0008469;  HP:0410198  MP:0005562     13.139874  0.300000   
2020913  HP_0000271;  HP:0002553  MP:0003761     13.833021  0.153846   
1361777  HP_0000271;  MP:0003761  HP:0002553     13.833021  0.153846   

         resnik_diff  
962688      9.199583  
632818      9.199583  
1327215     9.191583  
1326917     9.191583  
878712      9.191583  
...              ...  
1071039   -10.631039  
728809    -10.631039  
1069220   -10.631039  
2020913   -11.320644  
1361777   -11.320644  

[5833600 rows x 10 columns]

In [ ]:
intersected_scores['resnik_diff'].plot(kind='hist')

In [ ]:
intersected_scores.boxplot(column='resnik_diff')

In [8]:
# Find pairs +/- 1.5 iqr
q3 = np.quantile(intersected_scores['resnik_diff'], 0.75)
q1 = np.quantile(intersected_scores['resnik_diff'], 0.25)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
(lower, upper)

(-6.1562406229018185, -0.16051698226410993)

In [9]:
# Get pairs outside that range (lower than the lower bound, higher than the higher bound)
divergent_scores = intersected_scores[(intersected_scores['resnik_diff'] <= lower) | (intersected_scores['resnik_diff'] >= upper)]
divergent_scores

source_old destination_old  jaccard_old  resnik_score_old  \
962688   HP:0002607      MP:0003281     1.000000         14.359005   
632818   HP:0040183      MP:0003281     0.944444         14.359005   
1327215  HP:0012700      MP:0003281     0.529412         14.351005   
1326917  HP:0012701      MP:0003281     0.500000         14.351005   
878712   HP:0012700      MP:0003281     0.529412         14.351005   
...             ...             ...          ...               ...   
1071039  HP:0410188      MP:0005562     0.282051          2.508834   
728809   HP:0410198      MP:0005562     0.321429          2.508834   
1069220  HP:0410198      MP:0005562     0.321429          2.508834   
2020913  HP:0002553      MP:0003761     0.222222          2.512377   
1361777  HP:0002553      MP:0003761     0.222222          2.512377   

        subsumer_old      source destination  resnik_score   jaccard  \
962688   MP_0003281;  HP:0002607  MP:0003281      5.159422  0.300000   
632818   MP_0003281;  HP:0040183  MP:0003281      5.159422  0.272727   
1327215  HP_0012700;  HP:0012700  MP:0003281      5.159422  0.333333   
1326917  HP_0012700;  HP:0012701  MP:0003281      5.159422  0.300000   
878712   HP_0012700;  MP:0003281  HP:0012700      5.159422  0.333333   
...              ...         ...         ...           ...       ...   
1071039  MP_0008469;  HP:0410188  MP:0005562     13.139874  0.300000   
728809   MP_0008469;  MP:0005562  HP:0410198     13.139874  0.300000   
1069220  MP_0008469;  HP:0410198  MP:0005562     13.139874  0.300000   
2020913  HP_0000271;  HP:0002553  MP:0003761     13.833021  0.153846   
1361777  HP_0000271;  MP:0003761  HP:0002553     13.833021  0.153846   

         resnik_diff  
962688      9.199583  
632818      9.199583  
1327215     9.191583  
1326917     9.191583  
878712      9.191583  
...              ...  
1071039   -10.631039  
728809    -10.631039  
1069220   -10.631039  
2020913   -11.320644  
1361777   -11.320644  

[211443 rows x 10 columns]

In [ ]:
divergent_scores[(divergent_scores['resnik_diff'] <= lower)]

Overall, most pairs have higher Resnik similarity with new methods than with older ones. More than 211 thousand pairs meet our cutoff.

The top 5 differences in Resnik score are for pairs involving MP:0003281 (fecal incontinence). The top HP hit in the old scores is for HP:0002607 (bowel incontinence) and even includes the synonym "Fecal incontinence", so that appears to be a very good match. The low score in the new set may reflect how the ancestors of these terms are otherwise dissimilar: 
In HP, it looks like
```
All
 Phenotypic abnormality
  Abnormality of the digestive system
   Abnormality of digestive system physiology
    Abnormal large intestine physiology
     Bowel incontinence
  Constitutional symptom
   Impairment of activities of daily living
    Impaired continence
     Bowel incontinence
```
While in MP, we have
```
mammalian phenotype
 digestive/alimentary phenotype
  abnormal digestive system physiology
   abnormal defecation
    fecal incontinence
```

The other HP terms in this top 5 are HP:0040183 (Encopresis; a child of HP:0002607), HP:0012700 (Abnormal large intestine physiology; see above), HP:0012701 (Bowel urgency; a child of HP:0012700), and HP:0012700 (that's a duplicate - shouldn't be there).

The *lowest* differences (i.e., the pairs where Resnik was previously very low and is now very high) involve HP:0410188 (Decreased glucose-6-phosphate dehydrogenase level in red blood cells), MP:0005562 (decreased mean corpuscular hemoglobin),  HP:0002553 (Highly arched eyebrow), and MP:0003761 (arched palate). Several of these are duplicates. In the latter pair, the lexical similarity of "arched" is a clue, as the path between these terms involves UPHENO:0041110 (arched anatomical entity). Related to PATO:0001594 (arched), this term is useful for describing a type of similarity but isn't meaningful for phenotype similarity, at least in this case. For the former pair of HP:0410188 vs MP:0005562, the phenotypes appear similar on the basis of involving blood. The path between them crosses through UPHENO:0034498 (decreased level of chemical entity in erythrocyte)

In [ ]:
# Let's see a few more of those pairs.
divergent_scores.tail(20)

HP:0002227 (White eyelashes) vs MP:0006242 (white pupil) is a pair with physical similarity in both anatomy and attribute. It has a relatively high score of 13.427556 in the new method, as does HP:0002226 (White eyebrow).

HP:0006611 (Decreased number of sternal ossification centers) vs MP:0010869 (decreased bone trabecula number): both involve bone, certainly, though the high similarity score may be misrepresenting what is otherwise a dissimilar phenotype.

## OAK and Graph viz

Time to visualize, courtesy of OAK, obographviz, and KG-Phenio.

In [ ]:
# Set up the alias first for notebook purposes
%alias phenio runoak -i sqlite:obo:phenio

In [ ]:
phenio info "HP:0006611"

In [ ]:
phenio viz -p i,p "HP:0006611" "MP:0010869" -o output2.png --no-view

## Retrieve all UPHENO MRCAs for divergent HP-MP pairs

In [32]:
div_pairs = list(zip(divergent_scores["source"], divergent_scores["destination"]))

('HP:0002607', 'MP:0003281')
('HP:0040183', 'MP:0003281')
('HP:0012700', 'MP:0003281')
('HP:0012701', 'MP:0003281')
('MP:0003281', 'HP:0012700')
('HP:0012702', 'MP:0003281')
('HP:0012423', 'MP:0003281')
('HP:0100819', 'MP:0003319')
('HP:0033279', 'MP:0003319')
('HP:0100744', 'MP:0030883')


In [11]:
test1 = "HP:0006611"
test2 = "MP:0010869"

The following will retrieve all shortest paths for all the divergent pairs. Each requires processing once complete to:
* Remove header lines
* Select only the UPHENO part of the path
* Find the UPHENO midpoint for each

Doing the whole set this way takes intractably long, so let's try the last 1000.

In [35]:
all_spaths = {}
count = 1000

In [36]:
%%time
for pair in div_pairs[0:count]:
    spath = !runoak -i sqlite:obo:phenio paths -p i,p {pair[0]} --target {pair[1]} --no-autolabel
    spath = ((spath[1:])[0]).replace("[", "").replace("]", "").replace("\t", ",").replace("\'","").replace(" ","")
    spath = spath.split(",")

    spath_upheno = []
    for node in spath:
        if node.startswith("UPHENO"):
            spath_upheno.append(node)

    spath_upheno_mrca = spath_upheno[len(spath_upheno)//2]        

    all_spaths[pair] = spath_upheno_mrca

all_spaths

CPU times: user 1.4 s, sys: 6.63 s, total: 8.03 s
Wall time: 47min 44s


{('HP:0002607', 'MP:0003281'): 'UPHENO:0002320',
 ('HP:0040183', 'MP:0003281'): 'UPHENO:0002320',
 ('HP:0012700', 'MP:0003281'): 'UPHENO:0002320',
 ('HP:0012701', 'MP:0003281'): 'UPHENO:0002320',
 ('MP:0003281', 'HP:0012700'): 'UPHENO:0002332',
 ('HP:0012702', 'MP:0003281'): 'UPHENO:0002320',
 ('HP:0012423', 'MP:0003281'): 'UPHENO:0002320',
 ('HP:0100819', 'MP:0003319'): 'UPHENO:0019528',
 ('HP:0033279', 'MP:0003319'): 'UPHENO:0019528',
 ('HP:0100744', 'MP:0030883'): 'UPHENO:0020736',
 ('MP:0012399', 'HP:0011937'): 'UPHENO:0081318',
 ('HP:0008755', 'MP:0031308'): 'UPHENO:0004536',
 ('HP:0002786', 'MP:0013753'): 'UPHENO:0004536',
 ('HP:0033990', 'MP:0013753'): 'UPHENO:0004536',
 ('HP:0033979', 'MP:0013753'): 'UPHENO:0004536',
 ('HP:0033979', 'MP:0031308'): 'UPHENO:0004536',
 ('HP:0002786', 'MP:0031308'): 'UPHENO:0004536',
 ('HP:0008755', 'MP:0013753'): 'UPHENO:0004536',
 ('HP:0002779', 'MP:0013753'): 'UPHENO:0004536',
 ('HP:0033990', 'MP:0031308'): 'UPHENO:0004536',
 ('HP:0002779', 'MP:

In [38]:
with open("upheno_paths.tsv", 'w') as outfile:
    for spath in all_spaths:
        outfile.write(f'{spath[0]}\t{spath[1]}\t{all_spaths[spath]}\n')